File: utils/onebot.py
import os

import aiohttp
from pydantic import BaseModel

from dependencies.verify_chat_info import ChatInfo

ONEBOT_API_URL = os.environ.get("ONEBOT_API_URL")
if not ONEBOT_API_URL:
    raise ValueError("ONEBOT_API_URL environment variable is not set.")

ONEBOT_ACCESS_TOKEN = os.environ.get("ONEBOT_ACCESS_TOKEN")


class SendMessageResponse(BaseModel):
    status: str
    retcode: int
    message: str


async def send_message(chat_info: ChatInfo, message: str):
    if chat_info.chat_type == "group":
        endpoint = "/send_group_msg"
        payload = {
            "group_id": chat_info.chat_number,
            "message": [{"type": "text", "data": {"text": message}}],
        }
    elif chat_info.chat_type == "private":
        endpoint = "/send_private_msg"
        payload = {
            "user_id": chat_info.chat_number,
            "message": [{"type": "text", "data": {"text": message}}],
        }
    else:
        raise ValueError("Invalid chat type. Must be 'group' or 'private'.")

    async with aiohttp.ClientSession() as session:
        headers = {"Content-Type": "application/json"}
        if ONEBOT_ACCESS_TOKEN:
            headers["Authorization"] = f"Bearer {ONEBOT_ACCESS_TOKEN}"

        async with session.post(
            f"{ONEBOT_API_URL}{endpoint}",
            json=payload,
            headers=headers,
        ) as response:
            if response.status != 200:
                return SendMessageResponse(
                    status="error",
                    retcode=response.status,
                    message="Failed to send message",
                )
            else:
                response_data = await response.json()
                return SendMessageResponse(
                    status=response_data.get("status"),
                    retcode=response_data.get("retcode"),
                    message=response_data.get("message"),
                )

File: src/lib/onebot.ts
import { ONEBOT_ACCESS_TOKEN, ONEBOT_API_URL } from "./env";

interface ChatInfo {
  chatType: string;
  chatNumber: string;
}

const endpointConfig = {
  group: {
    endpoint: "/send_group_msg",
    idField: "group_id",
  },
  private: {
    endpoint: "/send_private_msg",
    idField: "user_id",
  },
};

export const sendMessage = async (
  chatInfo: ChatInfo,
  message: string
): Promise<{
  error: string | null;
}> => {
  if (isNaN(Number(chatInfo.chatNumber))) {
    return { error: "Chat number must be a valid number" };
  }

  const config =
    endpointConfig[chatInfo.chatType as keyof typeof endpointConfig];
  if (!config) {
    return { error: "Invalid chat type" };
  }

  const payload = {
    [config.idField]: chatInfo.chatNumber,
    message: [{ type: "text", data: { text: message } }],
  };

  const headers: Record<string, string> = {
    "Content-Type": "application/json",
    ...(ONEBOT_ACCESS_TOKEN && {
      Authorization: `Bearer ${ONEBOT_ACCESS_TOKEN}`,
    }),
  };

  try {
    const response = await fetch(`${ONEBOT_API_URL}${config.endpoint}`, {
      method: "POST",
      headers: headers,
      body: JSON.stringify(payload),
    });

    const data = await response.json();

    if (!response.ok) {
      return { error: data.message || "Failed to send message" };
    }

    if (data.status != "ok") {
      return { error: data.message || "Failed to send message" };
    }

    return { error: null };
  } catch (error) {
    return {
      error: error instanceof Error ? error.message : "Unknown error occurred",
    };
  }
};
